rs/ethereum/cketh/minter/src/numeric/range.rs

```diff
@@ -0,0 +1,258 @@
+        self,
+        mid: T,
+    ) -> (Option<BlockRangeInclusive>, Option<BlockRangeInclusive>) {
+        let mid = mid.into();
+        if &mid <= self.as_ref().start() {
+            (None, Some(self))
+        } else if &mid > self.as_ref().end() {
+            (Some(self), None)
+        } else {
+            //0 <= start < mid <= end
+            let (start, end) = self.0.into_inner();
+            let left = BlockRangeInclusive::new(
+                start,
+                mid.checked_decrement()
+                    .expect("BUG: mid is strictly positive"),
+            );
+            let right = BlockRangeInclusive::new(mid, end);
+            (Some(left), Some(right))
+        }
+    }
+
+    /// Partition a block range into two non-overlapping ranges at the midpoint,
+    /// which is included in the first partition.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use ic_cketh_minter::numeric::BlockRangeInclusive;
+    ///
+    /// let block_range = BlockRangeInclusive::from(1..=1_u8);
+    /// let (left, right) = block_range.clone().partition_into_halves();
+    /// assert_eq!(left, Some(block_range));
+    /// assert_eq!(right, None);
+    ///
+    /// let block_range = BlockRangeInclusive::from(1..=2_u8);
+    /// let (left, right) = block_range.partition_into_halves();
+    /// assert_eq!(left, Some(BlockRangeInclusive::from(1..=1_u8)));
+    /// assert_eq!(right, Some(BlockRangeInclusive::from(2..=2_u8)));
+    ///
+    /// let block_range = BlockRangeInclusive::from(1..=3_u8);
+    /// let (left, right) = block_range.partition_into_halves();
+    /// assert_eq!(left, Some(BlockRangeInclusive::from(1..=2_u8)));
+    /// assert_eq!(right, Some(BlockRangeInclusive::from(3..=3_u8)));
+    ///
+    /// let block_range = BlockRangeInclusive::from(1..=4_u8);
+    /// let (left, right) = block_range.partition_into_halves();
+    /// assert_eq!(left, Some(BlockRangeInclusive::from(1..=2_u8)));
+    /// assert_eq!(right, Some(BlockRangeInclusive::from(3..=4_u8)));
+    /// ```
+    ///
+    pub fn partition_into_halves(
+        self,
+    ) -> (Option<BlockRangeInclusive>, Option<BlockRangeInclusive>) {
+        if self.as_ref().start() >= self.as_ref().end() {
+            return (Some(self), None);
+        }
+        // start < end <= BlockNumber::MAX
+        let mid = self
+            .as_ref()
+            .start()
+            .checked_add(
+                self.as_ref()
+                    .end()
+                    .checked_sub(*self.as_ref().start())
+                    .expect("unreachable: end > start")
+                    .div_by_two(),
+            )
+            // mid == end - (delta - delta/2) < end,
+            // where delta = end - start > 0
+            .expect("unreachable: mid < end")
+            .checked_increment()
+            .expect("unreachable: mid <= end");
+        self.partition_at_checked(mid)
+    }
+
+    /// Join two block ranges into a single range that spans both ranges.
```

Comment:
If there is a gap between the two ranges, the gap will also be part of the result range. Is this desirable?

Reply:
I think the name is fine as long as the description captures the semantics.